### ISSUE TITLE
Performance regression on NPU after #154

### ISSUE THREAD

**Issue by A1waysBeenHere** [ByteDance-Seed/VeOmni#209]
Qwen3-32B SFT performance regressed by 6x on NPU 
Take Qwen3 as a example.
#154 patched some code on Qwen3ForCausalLM.forward to keep cu_seq_lens_x on CPU in NPU case, but which lead to performance regression. A lot communication cannot be overlapped base on profiling. 
I am not sure if `.cpu()` Qwen3ForCausalLM.forward lead to some sync op. And it is all good once I moved the `.cpu()` to Qwen3Attention.forward.

And I get #199 to patch the `.cpu()` code in Attention class only.

PS: 
We noticed main branch code now, in my case, the memory usage will increase about 4 gigabytes, and which also leads a huge time-consuming in empty_tensor, we can`export MULTI_STREAM_MEMORY_REUSE=2` `export PYTORCH_NPU_ALLOC_CONF=expandable_segments:True`  to optimize memory usage and avoid this problem. But still confused about why #154 get extra 4 gigabytes in memory usage

**Comment by Luosuu** [comment]
Thank you for your investigation. cc @wey-code @piyifan123

**Comment by Crystal-jiang** [comment]
ascend

**Comment by wey-code** [comment]
@A1waysBeenHere 
Thanks for the contribution～

In our previous experiments on Qwen models, we did not observe similar extra synchronization being triggered by this change. From a theoretical perspective, the modification introduced in the referenced MR should also be unrelated to the observed increase in NPU memory usage.

To help us better understand and analyze the performance regression and memory behavior, could you please provide the relevant profiling data (e.g., profiler traces or logs)? That would be very helpful for further investigation.

Thanks again for your efforts and support.

**Comment by piyifan123** [comment]
+1 it would be nice to see a trace (you can use veomni's enable_profiling function). We don't observe this issue in our cases when training qwen3 dense models. The particular PR you refers to actually reduce the per layer attention cu seqlens GPU to CPU sync which was a huge performance degrader.

**Comment by A1waysBeenHere** [comment]
> [@A1waysBeenHere](https://github.com/A1waysBeenHere) Thanks for the contribution～
> 
> In our previous experiments on Qwen models, we did not observe similar extra synchronization being triggered by this change. From a theoretical perspective, the modification introduced in the referenced MR should also be unrelated to the observed increase in NPU memory usage.
> 
> To help us better understand and analyze the performance regression and memory behavior, could you please provide the relevant profiling data (e.g., profiler traces or logs)? That would be very helpful for further investigation.
> 
> Thanks again for your efforts and support.

Will get profiling soon

**Comment by A1waysBeenHere** [comment]
Cannot get profiling file for some reasons, but I got torch memory snapshot. I noticed that when we have `cu_seq_len_q.cpu()`,  compete layers accumulated and cannot be resharded in time, it's where the extra memory usage from.

Maybe synchronization triggered by .cpu() in `def forward` collapse the workflow (layer X unshard -> layer X compute/layer X + 1 unshard -> layer X reshard)

abnormal case: There are 11 layers at the end of the backward phase at the same time. Similar situations exist in both forward and backward phase.


<img width="2113" height="787" alt="Image" src="https://github.com/user-attachments/assets/46db7035-fed3-4973-91fb-2ea2bb53a013" />

normal case: 

<img width="2112" height="796" alt="Image" src="https://github.com/user-attachments/assets/1c1cf592-f1a0-45cf-99d0-ed53365eb55d" />

@piyifan123 @wey-code

**Comment by onehaitao** [comment]
I think `.cpu()` operation that caused stream synchronization, so the memory problem was alleviated. If there is no `.cpu()` operation, delayed memory reuse can lead to memory stacking and push up memory.

**Comment by wey-code** [comment]
> Cannot get profiling file for some reasons, but I got torch memory snapshot. I noticed that when we have `cu_seq_len_q.cpu()`, compete layers accumulated and cannot be resharded in time, it's where the extra memory usage from.
> 
> Maybe synchronization triggered by .cpu() in `def forward` collapse the workflow (layer X unshard -> layer X compute/layer X + 1 unshard -> layer X reshard)
> 
> abnormal case: There are 11 layers at the end of the backward phase at the same time. Similar situations exist in both forward and backward phase.
> 
> <img alt="Image" width="2000" height="787" src="https://private-user-images.githubusercontent.com/71289623/530599357-46db7035-fed3-4973-91fb-2ea2bb53a013.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc1MDExNTQsIm5iZiI6MTc2NzUwMDg1NCwicGF0aCI6Ii83MTI4OTYyMy81MzA1OTkzNTctNDZkYjcwMzUtZmVkMy00OTczLTkxZmItMmVhMmJiNTNhMDEzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTA0VDA0MjczNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2YTgxZDExNDBhMzY5ZDBhYzJlM2M1YTBlNjdiMTI2YTg1YmU2YzQyMDVhZjJlNTFjNDBhM2E0MTdmNjkzNDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XJcBz9FGZj4YHvk4LOCHHEMCkrPnMMqF_8jIM2_W7uA">
> normal case:
> 
> <img alt="Image" width="2000" height="796" src="https://private-user-images.githubusercontent.com/71289623/530599729-1c1cf592-f1a0-45cf-99d0-ed53365eb55d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc1MDExNTQsIm5iZiI6MTc2NzUwMDg1NCwicGF0aCI6Ii83MTI4OTYyMy81MzA1OTk3MjktMWMxY2Y1OTItZjFhMC00NWNmLTk5ZDAtZWQ1MzM2NWViNTVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTA0VDA0MjczNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMxYzdjODQzYzQwZjU5NDU4ZjU1ZWVjODU4NDBhMjg5MGM5OWY3ODg2MzY5MGVlYWVkNzdmNTAxODJmMzdkNTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.m3pN253Xmk8ryH0nU_xRYxAmj14OYPlnhDtUVyoHKh0">
> [@piyifan123](https://github.com/piyifan123) [@wey-code](https://github.com/wey-code)

This seems related to the FSDP implementation. Could you provide the versions of torch, torch_npu, CANN etc. that you're using?

**Comment by A1waysBeenHere** [comment]
> > Cannot get profiling file for some reasons, but I got torch memory snapshot. I noticed that when we have `cu_seq_len_q.cpu()`, compete layers accumulated and cannot be resharded in time, it's where the extra memory usage from.
> > Maybe synchronization triggered by .cpu() in `def forward` collapse the workflow (layer X unshard -> layer X compute/layer X + 1 unshard -> layer X reshard)
> > abnormal case: There are 11 layers at the end of the backward phase at the same time. Similar situations exist in both forward and backward phase.
> > <img alt="Image" width="2000" height="787" src="https://private-user-images.githubusercontent.com/71289623/530599357-46db7035-fed3-4973-91fb-2ea2bb53a013.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc1MDExNTQsIm5iZiI6MTc2NzUwMDg1NCwicGF0aCI6Ii83MTI4OTYyMy81MzA1OTkzNTctNDZkYjcwMzUtZmVkMy00OTczLTkxZmItMmVhMmJiNTNhMDEzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTA0VDA0MjczNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2YTgxZDExNDBhMzY5ZDBhYzJlM2M1YTBlNjdiMTI2YTg1YmU2YzQyMDVhZjJlNTFjNDBhM2E0MTdmNjkzNDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XJcBz9FGZj4YHvk4LOCHHEMCkrPnMMqF_8jIM2_W7uA">
> > normal case:
> > <img alt="Image" width="2000" height="796" src="https://private-user-images.githubusercontent.com/71289623/530599729-1c1cf592-f1a0-45cf-99d0-ed53365eb55d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njc1MDExNTQsIm5iZiI6MTc2NzUwMDg1NCwicGF0aCI6Ii83MTI4OTYyMy81MzA1OTk3MjktMWMxY2Y1OTItZjFhMC00NWNmLTk5ZDAtZWQ1MzM2NWViNTVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTA0VDA0MjczNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMxYzdjODQzYzQwZjU5NDU4ZjU1ZWVjODU4NDBhMjg5MGM5OWY3ODg2MzY5MGVlYWVkNzdmNTAxODJmMzdkNTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.m3pN253Xmk8ryH0nU_xRYxAmj14OYPlnhDtUVyoHKh0">
> > [[@piyifan123](https://github.com/piyifan123)](https://github.com/piyifan123) [[@wey-code](https://github.com/wey-code)](https://github.com/wey-code)
> 
> This seems related to the FSDP implementation. Could you provide the versions of torch, torch_npu, CANN etc. that you're using?

torch 2.7.1
torch_npu 2.7.1.dev20251023
CANN 8.2RC1
